README.md
# NAME

Short::URL - Encodes and Decodes short urls by using Bijection

# SYNOPSIS

       use Short::URL;
       
       #use default alphabet
       my $su = Short::URL->new;

       my $encoded = $su->encode(10000);
       my $short_url = "http://www.example.com/$encoded";

       print "Encoded: $encoded\n"; #prints cP6
       print "Short url: $short_url\n"; #prints http://www.example.com/cP6

       my $decoded = $su->decode('cP6');

       print "Decoded: $decoded\n"; #prints 10000

       #set new alphabet
       $su->alphabet([qw/1 2 3 a b c/]);

       #or when you create the Short::URL object
       my $su = Short::URL->new(alphabet => [qw/1 2 3 a b c/]);

       #have an 'offset'. Meaning if you pass in integer $i, you get unique string for $i + offset
       $su->offset(10000);

       #or
       my $su = Short::URL->new(offset => 10000);

       my $encoded_with_offset = $su->encode(0);

       print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6'

       my $decoded_with_offset = $su->decode('cP6');

       print "Decoded with offset: $decoded_with_offset\n"; #prints 0

       #use shuffled alphabet
       $su->use_shuffled_alphabet(1);

       #or
       my $su = Short::URL->new(use_shuffled_alphabet => 1);

       my $encoded_with_shuffled_alphabet = $su->encode(10000);

       print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

       my $decoded_with_shuffled_alphabet = $su->decode('dlu');

       print "Decoded with shuffled alphabet: $decoded_with_shuffled_alphabet"; #prints 10000

       #use no naughty words
       $su->no_vowels(1);

       #or
       my $su = Short::URL->new(no_vowels => 1);

       my $clean_encoded = $su->encode(7465182);

       print "My clean word: $clean_encoded\n"; #prints 'ccgZn', instead of 'F_ck', where '_' is replaced with 'u'
    

# DESCRIPTION

Short::URL can be used to help generate short, unique character string urls. It uses [Bijection](http://en.wikipedia.org/wiki/Bijection) to
create a one-to-one mapping from integers to strings over your alphabet, and from strings over your alphabet back to the original integer. An integer
primary key in your database would be a good example of an integer you could use to generate a unique character string that maps uniquely to that row.

# METHODS

## alphabet

    $su->alphabet([qw/1 2 3 a b c/]);

The alphabet that will be used for creating strings when mapping an integer to a string. The default alphabet is:

    [qw/a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

All lower case letters, upper case letters, and digits 0-9.

## encode

    my $encoded = $su->encode(1);

Takes an integer and encodes it into a unique string over your alphabet.

## decode

    my $decoded = $su->decode('b');

Takes a string made from your alphabet and returns the corresponding integer that it maps to.

## offset

    $su->offset(10000);
    my $encoded_with_offset = $su->encode(0);
     
    print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6' instead of 'a'
     
    my $decoded_with_offset = $su->decode('cP6');
     
    print "Decoded with offset: $decoded_with_offset\n"; #prints 0 instead of 10000

["offset"](#offset) allows you to shift all integers that you encode forward by a certain integer. Likewise, whenever you are decoding strings it will
return the integer for that string by substracting your offset from what the original number would have been. Offset can be useful if you don't want people
to know you have a small number of items by starting with a larger offset, or if you want them to have less of an idea what ids really correspond to what strings.
When used in combination with ["use\_shuffled\_alphabet"](#use_shuffled_alphabet), it is a lot harder to track what string would correspond
to what id in your database, or how many ids you have in total. 

## use\_shuffled\_alphabet

    $su->use_shuffled_alphabet(1);

    my $encoded_with_shuffled_alphabet = $su->encode(10000);

    print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

Setting ["use\_shuffled\_alphabet"](#use_shuffled_alphabet) to 1 means that instead of using the more common alphabet used for this algorithm that is stored in ["alphabet"](#alphabet), [Short::URL](https://metacpan.org/pod/Short::URL) will use a shuffled alphabet.
Note, this shuffled alphabet will be the same every time, so encoding and decoding will work even with the use\_shuffled\_alphabet set to 1 even in 
different sessions of using [Short::URL](https://metacpan.org/pod/Short::URL). This can be 
useful if for some reason you don't want people to know how many ids you have in your database. With the more standard alphabet it's clear when you are going
up by one between strings, and thus also how many ids you have. When used in combination with ["offset"](#offset), it is a lot harder to track what string would correspond
to what id in your databse, or how many ids you have in total. Below is the shuffled alphabet that is used:

    [qw/G w d A t H J 0 P o W C 6 3 y K 8 L u 7 X E O a e q 1 9 D c F x Z 5 M T N l z r i s j h B 4 b I f V Y Q g n S 2 m U k p v R/]

This is just a shuffled version of ["alphabet"](#alphabet). The default for ["use\_shuffled\_alphabet"](#use_shuffled_alphabet) is undef.

## no\_vowels

As pointed out [here](http://stackoverflow.com/questions/742013/how-to-code-a-url-shortener/742047#comment25208796_742047), there is the possibility of creating
naughty words. For instance:

    my $encoded = $su->encode(7465182);

    print "Encoded naughty: $encoded\n"; #prints F_ck, where '_' is replaced with 'u'

If you want to avoid this, you can set ["no\_vowels"](#no_vowels) to 1:

    $su->no_vowels(1);

    #or
    my $su = Short::URL->new(no_vowels => 1);

This uses a separate alphabet with no vowels in it (aeiou), so that no naughty words can be formed. If you also have ["use\_shuffled\_alphabet"](#use_shuffled_alphabet) set to 1, that will still be
respected. [Short::URL](https://metacpan.org/pod/Short::URL) will just use a clean version of the shuffled alphabet listed under ["use\_shuffled\_alphabet"](#use_shuffled_alphabet). Below are the two clean alphabets:

    #clean_alphabet
    [qw/b c d f g h j k l m n p q r s t v w x y z B C D F G H J K L M N P Q R S T V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

    #clean_shuffled_alphabet
    [qw/G w d t H J 0 P W C 6 3 y K 8 L 7 X q 1 9 D c F x Z 5 M T N l z r s j h B 4 b f V Y Q g n S 2 m k p v R/]

## croak\_on\_error

This method sets whether you want [Short::URL](https://metacpan.org/pod/Short::URL) to ["croak"](https://metacpan.org/pod/Carp) on an error.

    #enable
    $su->croak_on_error(1);

    #disable
    $su->croak_on_error(0);

Default is 1. If ["croak\_on\_error"](#croak_on_error) is set to false and there is an error, then ["encode"](#encode) or ["decode"](#decode) will return a negative number.

# AUTHOR

Adam Hopkins <srchulo@cpan.org>

# COPYRIGHT

Copyright 2019- Adam Hopkins

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

- [http://stackoverflow.com/a/742047/834140](http://stackoverflow.com/a/742047/834140)
- [https://gist.github.com/zumbojo/1073996](https://gist.github.com/zumbojo/1073996)


Changes
Revision history for Short::URL

{{$NEXT}}

0.015  2019-10-04 14:31:17 CDT
    - Fixed typo in documentation to Carp (@manwar).

0.014  2019-01-13 15:04:32 CST
    - Switched to using Dist::Milla for releases
    - Switched to Moo

0.013     2016-11-14 23:23:37-08:00 America/Los_Angeles
    Renamed Url.pm to URL.pm

0.012     2016-02-07 18:56:38-08:00 America/Los_Angeles
    Added the package version.

0.011     2014-08-18 11:55:49-05:00 America/Chicago
    Removed print and fixed "databse" typo

0.010     2014-08-03 11:48:15-07:00 America/Los_Angeles
    Added offset to documentation

0.009     2014-08-03 00:53:04-07:00 America/Los_Angeles
    Fixed some typos in Synopsis thanks to azawawi

0.008     2014-08-02 19:44:53-07:00 America/Los_Angeles
    Store alphabet_in_use instead of calling many times in loop

0.007     2014-07-27 12:07:31-07:00 America/Los_Angeles
    Removed pointles use List::Util line

0.006     2014-07-26 23:29:44-07:00 America/Los_Angeles
    Fixed issue where reverse was used instead of return in encode.

0.005     2014-07-26 22:59:48-07:00 America/Los_Angeles
    Changed no_naughties to no_vowels. Fixed issue where Template::Toolkit didn't handled returned reverse value properly.
    Also, added croak_on_error.

0.004     2014-07-26 18:24:52-07:00 America/Los_Angeles
    Added no_naughties to synopsis

0.003     2014-07-26 18:17:56-07:00 America/Los_Angeles
    Added no_naughties

0.002     2014-07-26 17:10:58-07:00 America/Los_Angeles
    Removed test.pl from distribution

0.001     2014-07-26 16:59:58-07:00 America/Los_Angeles
	Released to the world!


lib/Short/Url.pm
package Short::URL;
use strict;
use Mouse;
use Carp qw//;
#ABSTRACT: Encodes and Decodes short urls by using Bijection

has alphabet => (
    is => 'rw',
    isa => 'ArrayRef',
    default => sub {
        [qw/a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9/]
    },  
);

has clean_alphabet => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub {
        [qw/b c d f g h j k l m n p q r s t v w x y z B C D F G H J K L M N P Q R S T V W X Y Z 0 1 2 3 4 5 6 7 8 9/]
    },  
);

has shuffled_alphabet => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub {
        [qw/G w d A t H J 0 P o W C 6 3 y K 8 L u 7 X E O a e q 1 9 D c F x Z 5 M T N l z r i s j h B 4 b I f V Y Q g n S 2 m U k p v R/]
    },  
);

has clean_shuffled_alphabet => (
    is => 'ro',
    isa => 'ArrayRef',
    default => sub {
        [qw/G w d t H J 0 P W C 6 3 y K 8 L 7 X q 1 9 D c F x Z 5 M T N l z r s j h B 4 b f V Y Q g n S 2 m k p v R/]
    },  
);

has use_shuffled_alphabet => (
    is => 'rw',
    isa => 'Any',
    default => undef,
);

has no_vowels => (
    is => 'rw',
    isa => 'Any',
    default => undef,
);

has offset => (
    is => 'rw',
    isa => 'Int',
    default => 0,
);

has croak_on_error => (
    is => 'rw',
    isa => 'Any',
    default => 1,
);

sub alphabet_in_use { 
    my ($self) = @_;
    my $alphabet;

    if($self->no_vowels) {
        $alphabet = $self->use_shuffled_alphabet ? $self->clean_shuffled_alphabet : $self->clean_alphabet;
    }
    else {
        $alphabet = $self->use_shuffled_alphabet ? $self->shuffled_alphabet : $self->alphabet;
    }

    print "@$alphabet\n";
    return $alphabet;
}

sub encode { 
    my ($self, $i) = @_; 

    $i += $self->offset;

    if($i < 0) { 
        Carp::croak "integer to encode must be nonnegative" if $self->croak_on_error;
        return -1;
    }

    my $alphabet = $self->alphabet_in_use;
    return $alphabet->[0] if $i == 0;

    my $s = ''; 
    my $base = @{$alphabet};

    while($i > 0) {
        $s .= $alphabet->[$i % $base];
        $i = int($i / $base);
    }   

    my $reversed_string = reverse $s;
    return $reversed_string;
}

sub decode { 
    my ($self, $s) = @_;
    
    my $i = 0;
    my $alphabet = $self->alphabet_in_use;
    my $base = @{$alphabet};
    my $last_index = $#{$alphabet};
    
    for my $char (split //, $s) { 
        my ($index) = grep { $alphabet->[$_] eq $char } 0..$last_index;

        if(not defined($index)) {
            Carp::croak "invalid character $char in $s" if $self->croak_on_error;
            return -1;
        }
        
        $i = $i * $base + $index;
    }

    $i -= $self->offset;

    Carp::croak "invalid string $s for offset @{[$self->offset]}. produces negative int: $i" if $i < 0 and $self->croak_on_error;

    return $i;
}

1;

__END__ 

=head1 SYNOPSIS
 
    use Short::URL;
    
    #use default alphabet
    my $su = Short::URL->new;

    my $encoded = $su->encode(10000);
    my $short_url = "http://www.example.com/$encoded";

    print "Encoded: $encoded\n"; #prints cP6
    print "Short url: $short_url\n"; #prints http://www.example.com/cP6

    my $decoded = $su->decode('cP6');

    print "Decoded: $decoded\n"; #prints 10000

    #set new alphabet
    $su->alphabet([qw/1 2 3 a b c/]);

    #or when you create the Short::URL object
    my $su = Short::URL->new(alphabet => [qw/1 2 3 a b c/]);

    #have an 'offset'. Meaning if you pass in integer $i, you get unique string for $i + offset
    $su->offset(10000);

    #or
    my $su = Short::URL->new(offset => 10000);

    my $encoded_with_offset = $su->encode(0);

    print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6'

    my $decoded_with_offset = $su->decode('cP6');

    print "Decoded with offset: $decoded_with_offset\n"; #prints 0

    #use shuffled alphabet
    $su->use_shuffled_alphabet(1);

    #or
    my $su = Short::URL->new(use_shuffled_alphabet => 1);

    my $encoded_with_shuffled_alphabet = $su->encode(10000);

    print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

    my $decoded_with_shuffled_alphabet = $su->decode('dlu');

    print "Decoded with shuffled alphabet: $decoded_with_shuffled_alphabet"; #prints 10000

    #use no naughty words
    $su->no_vowels(1);

    #or
    my $su = Short::URL->new(no_vowels => 1);

    my $clean_encoded = $su->encode(7465182);

    print "My clean word: $clean_encoded\n"; #prints 'ccgZn', instead of 'F_ck', where '_' is replaced with 'u'
 
=head1 DESCRIPTION

Short::URL can be used to help generate short, unique character string urls. It uses L<Bijection|http://en.wikipedia.org/wiki/Bijection> to
create a one-to-one mapping from integers to strings over your alphabet, and from strings over your alphabet back to the original integer. An integer
primary key in your database would be a good example of an integer you could use to generate a unique character string that maps uniquely to that row.
 
=method alphabet
 
    $su->alphabet([qw/1 2 3 a b c/]);

The alphabet that will be used for creating strings when mapping an integer to a string. The default alphabet is:

   [qw/a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

All lower case letters, upper case letters, and digits 0-9.

=method encode

    my $encoded = $su->encode(1);

Takes an integer and encodes it into a unique string over your alphabet.

=method decode

    my $decoded = $su->decode('b');

Takes a string made from your alphabet and returns the corresponding integer that it maps to.

=method offset

    $su->offset(10000);
    my $encoded_with_offset = $su->encode(0);
     
    print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6' instead of 'a'
     
    my $decoded_with_offset = $su->decode('cP6');
     
    print "Decoded with offset: $decoded_with_offset\n"; #prints 0 instead of 10000

L</offset> allows you to shift all integers that you encode forward by a certain integer. Likewise, whenever you are decoding strings it will
return the integer for that string by substracting your offset from what the original number would have been. Offset can be useful if you don't want people
to know you have a small number of items by starting with a larger offset, or if you want them to have less of an idea what ids really correspond to what strings.
When used in combination with L</use_shuffled_alphabet>, it is a lot harder to track what string would correspond
to what id in your databse, or how many ids you have in total. 

=method use_shuffled_alphabet

    $su->use_shuffled_alphabet(1);

    my $encoded_with_shuffled_alphabet = $su->encode(10000);

    print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

Setting L</use_shuffled_alphabet> to 1 means that instead of using the more common alphabet used for this algorithm that is stored in L</alphabet>, L<Short::URL> will use a shuffled alphabet.
Note, this shuffled alphabet will be the same every time, so encoding and decoding will work even with the use_shuffled_alphabet set to 1 even in 
different sessions of using L<Short::URL>. This can be 
useful if for some reason you don't want people to know how many ids you have in your database. With the more standard alphabet it's clear when you are going
up by one between strings, and thus also how many ids you have. When used in combination with L</offset>, it is a lot harder to track what string would correspond
to what id in your databse, or how many ids you have in total. Below is the shuffled alphabet that is used:

    [qw/G w d A t H J 0 P o W C 6 3 y K 8 L u 7 X E O a e q 1 9 D c F x Z 5 M T N l z r i s j h B 4 b I f V Y Q g n S 2 m U k p v R/]

This is just a shuffled version of L</alphabet>. The default for L</use_shuffled_alphabet> is undef.

=method no_vowels

As pointed out L<here|http://stackoverflow.com/questions/742013/how-to-code-a-url-shortener/742047#comment25208796_742047>, there is the possibility of creating
naughty words. For instance:

    my $encoded = $su->encode(7465182);

    print "Encoded naughty: $encoded\n"; #prints F_ck, where '_' is replaced with 'u'

If you want to avoid this, you can set L</no_vowels> to 1:

    $su->no_vowels(1);

    #or
    my $su = Short::URL->new(no_vowels => 1);

This uses a separate alphabet with no vowels in it (aeiou), so that no naughty words can be formed. If you also have L</use_shuffled_alphabet> set to 1, that will still be
respected. L<Short::URL> will just use a clean version of the shuffled alphabet listed under L</use_shuffled_alphabet>. Below are the two clean alphabets:

    #clean_alphabet
    [qw/b c d f g h j k l m n p q r s t v w x y z B C D F G H J K L M N P Q R S T V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

    #clean_shuffled_alphabet
    [qw/G w d t H J 0 P W C 6 3 y K 8 L 7 X q 1 9 D c F x Z 5 M T N l z r s j h B 4 b f V Y Q g n S 2 m k p v R/]

=method croak_on_error

This method sets whether you want L<Short::URL> to L<Carp|/"croak"> on an error. 

    #enable
    $su->croak_on_error(1);

    #disable
    $su->croak_on_error(0);

Default is 1. If L</croak_on_error> is set to false and there is an error, then L</encode> or L</decode> will return a negative number.

=head1 SEE ALSO

=over 4

=item

L<http://stackoverflow.com/a/742047/834140>

=item

L<https://gist.github.com/zumbojo/1073996>

=back


lib/Short/URL.pm
package Short::URL;

use 5.008_005;
our $VERSION = '0.014';

use strictures 2;
use Moo;
use Types::Standard qw/ArrayRef Bool Int Str/;
use Carp qw//;

has alphabet => (
    is => 'rw',
    isa => ArrayRef[Str],
    default => sub {
        [qw/a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9/]
    },  
);

has clean_alphabet => (
    is => 'ro',
    isa => ArrayRef[Str],
    default => sub {
        [qw/b c d f g h j k l m n p q r s t v w x y z B C D F G H J K L M N P Q R S T V W X Y Z 0 1 2 3 4 5 6 7 8 9/]
    },  
);

has shuffled_alphabet => (
    is => 'ro',
    isa => ArrayRef[Str],
    default => sub {
        [qw/G w d A t H J 0 P o W C 6 3 y K 8 L u 7 X E O a e q 1 9 D c F x Z 5 M T N l z r i s j h B 4 b I f V Y Q g n S 2 m U k p v R/]
    },  
);

has clean_shuffled_alphabet => (
    is => 'ro',
    isa => ArrayRef[Str],
    default => sub {
        [qw/G w d t H J 0 P W C 6 3 y K 8 L 7 X q 1 9 D c F x Z 5 M T N l z r s j h B 4 b f V Y Q g n S 2 m k p v R/]
    },  
);

has use_shuffled_alphabet => (
    is => 'rw',
    isa => Bool,
    default => undef,
);

has no_vowels => (
    is => 'rw',
    isa => Bool,
    default => undef,
);

has offset => (
    is => 'rw',
    isa => Int,
    default => 0,
);

has croak_on_error => (
    is => 'rw',
    isa => Bool,
    default => 1,
);

sub alphabet_in_use { 
    my ($self) = @_;
    my $alphabet;

    if($self->no_vowels) {
        $alphabet = $self->use_shuffled_alphabet ? $self->clean_shuffled_alphabet : $self->clean_alphabet;
    }
    else {
        $alphabet = $self->use_shuffled_alphabet ? $self->shuffled_alphabet : $self->alphabet;
    }

    return $alphabet;
}

sub encode { 
    my ($self, $i) = @_; 

    $i += $self->offset;

    if($i < 0) { 
        Carp::croak "integer to encode must be nonnegative" if $self->croak_on_error;
        return -1;
    }

    my $alphabet = $self->alphabet_in_use;
    return $alphabet->[0] if $i == 0;

    my $s = ''; 
    my $base = @{$alphabet};

    while($i > 0) {
        $s .= $alphabet->[$i % $base];
        $i = int($i / $base);
    }   

    my $reversed_string = reverse $s;
    return $reversed_string;
}

sub decode { 
    my ($self, $s) = @_;
    
    my $i = 0;
    my $alphabet = $self->alphabet_in_use;
    my $base = @{$alphabet};
    my $last_index = $#{$alphabet};
    
    for my $char (split //, $s) { 
        my ($index) = grep { $alphabet->[$_] eq $char } 0..$last_index;

        if(not defined($index)) {
            Carp::croak "invalid character $char in $s" if $self->croak_on_error;
            return -1;
        }
        
        $i = $i * $base + $index;
    }

    $i -= $self->offset;

    Carp::croak "invalid string $s for offset @{[$self->offset]}. produces negative int: $i" if $i < 0 and $self->croak_on_error;

    return $i;
}

1;

__END__ 

=head1 NAME

Short::URL - Encodes and Decodes short urls by using Bijection

=head1 SYNOPSIS
 
    use Short::URL;
    
    #use default alphabet
    my $su = Short::URL->new;

    my $encoded = $su->encode(10000);
    my $short_url = "http://www.example.com/$encoded";

    print "Encoded: $encoded\n"; #prints cP6
    print "Short url: $short_url\n"; #prints http://www.example.com/cP6

    my $decoded = $su->decode('cP6');

    print "Decoded: $decoded\n"; #prints 10000

    #set new alphabet
    $su->alphabet([qw/1 2 3 a b c/]);

    #or when you create the Short::URL object
    my $su = Short::URL->new(alphabet => [qw/1 2 3 a b c/]);

    #have an 'offset'. Meaning if you pass in integer $i, you get unique string for $i + offset
    $su->offset(10000);

    #or
    my $su = Short::URL->new(offset => 10000);

    my $encoded_with_offset = $su->encode(0);

    print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6'

    my $decoded_with_offset = $su->decode('cP6');

    print "Decoded with offset: $decoded_with_offset\n"; #prints 0

    #use shuffled alphabet
    $su->use_shuffled_alphabet(1);

    #or
    my $su = Short::URL->new(use_shuffled_alphabet => 1);

    my $encoded_with_shuffled_alphabet = $su->encode(10000);

    print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

    my $decoded_with_shuffled_alphabet = $su->decode('dlu');

    print "Decoded with shuffled alphabet: $decoded_with_shuffled_alphabet"; #prints 10000

    #use no naughty words
    $su->no_vowels(1);

    #or
    my $su = Short::URL->new(no_vowels => 1);

    my $clean_encoded = $su->encode(7465182);

    print "My clean word: $clean_encoded\n"; #prints 'ccgZn', instead of 'F_ck', where '_' is replaced with 'u'
 
=head1 DESCRIPTION

Short::URL can be used to help generate short, unique character string urls. It uses L<Bijection|http://en.wikipedia.org/wiki/Bijection> to
create a one-to-one mapping from integers to strings over your alphabet, and from strings over your alphabet back to the original integer. An integer
primary key in your database would be a good example of an integer you could use to generate a unique character string that maps uniquely to that row.
 
=head1 METHODS

=head2 alphabet
 
    $su->alphabet([qw/1 2 3 a b c/]);

The alphabet that will be used for creating strings when mapping an integer to a string. The default alphabet is:

   [qw/a b c d e f g h i j k l m n o p q r s t u v w x y z A B C D E F G H I J K L M N O P Q R S T U V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

All lower case letters, upper case letters, and digits 0-9.

=head2 encode

    my $encoded = $su->encode(1);

Takes an integer and encodes it into a unique string over your alphabet.

=head2 decode

    my $decoded = $su->decode('b');

Takes a string made from your alphabet and returns the corresponding integer that it maps to.

=head2 offset

    $su->offset(10000);
    my $encoded_with_offset = $su->encode(0);
     
    print "Encoded with offset: $encoded_with_offset\n"; #prints unique string for 10000, 'cP6' instead of 'a'
     
    my $decoded_with_offset = $su->decode('cP6');
     
    print "Decoded with offset: $decoded_with_offset\n"; #prints 0 instead of 10000

L</offset> allows you to shift all integers that you encode forward by a certain integer. Likewise, whenever you are decoding strings it will
return the integer for that string by substracting your offset from what the original number would have been. Offset can be useful if you don't want people
to know you have a small number of items by starting with a larger offset, or if you want them to have less of an idea what ids really correspond to what strings.
When used in combination with L</use_shuffled_alphabet>, it is a lot harder to track what string would correspond
to what id in your database, or how many ids you have in total. 

=head2 use_shuffled_alphabet

    $su->use_shuffled_alphabet(1);

    my $encoded_with_shuffled_alphabet = $su->encode(10000);

    print "Encoded with shuffled alphabet: $encoded_with_shuffled_alphabet"; #prints 'dlu'

Setting L</use_shuffled_alphabet> to 1 means that instead of using the more common alphabet used for this algorithm that is stored in L</alphabet>, L<Short::URL> will use a shuffled alphabet.
Note, this shuffled alphabet will be the same every time, so encoding and decoding will work even with the use_shuffled_alphabet set to 1 even in 
different sessions of using L<Short::URL>. This can be 
useful if for some reason you don't want people to know how many ids you have in your database. With the more standard alphabet it's clear when you are going
up by one between strings, and thus also how many ids you have. When used in combination with L</offset>, it is a lot harder to track what string would correspond
to what id in your databse, or how many ids you have in total. Below is the shuffled alphabet that is used:

    [qw/G w d A t H J 0 P o W C 6 3 y K 8 L u 7 X E O a e q 1 9 D c F x Z 5 M T N l z r i s j h B 4 b I f V Y Q g n S 2 m U k p v R/]

This is just a shuffled version of L</alphabet>. The default for L</use_shuffled_alphabet> is undef.

=head2 no_vowels

As pointed out L<here|http://stackoverflow.com/questions/742013/how-to-code-a-url-shortener/742047#comment25208796_742047>, there is the possibility of creating
naughty words. For instance:

    my $encoded = $su->encode(7465182);

    print "Encoded naughty: $encoded\n"; #prints F_ck, where '_' is replaced with 'u'

If you want to avoid this, you can set L</no_vowels> to 1:

    $su->no_vowels(1);

    #or
    my $su = Short::URL->new(no_vowels => 1);

This uses a separate alphabet with no vowels in it (aeiou), so that no naughty words can be formed. If you also have L</use_shuffled_alphabet> set to 1, that will still be
respected. L<Short::URL> will just use a clean version of the shuffled alphabet listed under L</use_shuffled_alphabet>. Below are the two clean alphabets:

    #clean_alphabet
    [qw/b c d f g h j k l m n p q r s t v w x y z B C D F G H J K L M N P Q R S T V W X Y Z 0 1 2 3 4 5 6 7 8 9/]

    #clean_shuffled_alphabet
    [qw/G w d t H J 0 P W C 6 3 y K 8 L 7 X q 1 9 D c F x Z 5 M T N l z r s j h B 4 b f V Y Q g n S 2 m k p v R/]

=head2 croak_on_error

This method sets whether you want L<Short::URL> to L<"croak"|Carp> on an error.

    #enable
    $su->croak_on_error(1);

    #disable
    $su->croak_on_error(0);

Default is 1. If L</croak_on_error> is set to false and there is an error, then L</encode> or L</decode> will return a negative number.

=head1 AUTHOR

Adam Hopkins E<lt>srchulo@cpan.orgE<gt>

=head1 COPYRIGHT

Copyright 2019- Adam Hopkins

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

=over 4

=item

L<http://stackoverflow.com/a/742047/834140>

=item

L<https://gist.github.com/zumbojo/1073996>

=back
